sdk/core/Azure.Core.TestFramework/src/TestEnvironment.cs

```diff
@@ -229,36 +232,158 @@ protected virtual ValueTask<bool> IsEnvironmentReadyAsync()
+                return;
+            }
+
+            // determine whether it is even possible to talk to the management endpoint
+            string resourceGroup = GetOptionalVariable("RESOURCE_GROUP");
+            if (resourceGroup == null)
+            {
+                return;
+            }
+
+            string subscription = GetOptionalVariable("SUBSCRIPTION_ID");
+            if (subscription == null)
+            {
+                return;
+            }
+
+            string tenantId = GetOptionalVariable("TENANT_ID");
+            string clientId = GetOptionalVariable("CLIENT_ID");
+            string clientSecret = GetOptionalVariable("CLIENT_SECRET");
+            string authorityHost = GetOptionalVariable("AZURE_AUTHORITY_HOST");
+
+            if (tenantId == null || clientId == null || clientSecret == null || authorityHost == null)
+            {
+                return;
+            }
+
+            // intentionally not using the Credential property as we don't want to throw if the env vars are not available, and we want to allow this to vary per environment.
+            var credential = new ClientSecretCredential(
+                    tenantId,
+                    clientId,
+                    clientSecret,
+                    new ClientSecretCredentialOptions()
+                    {
+                        AuthorityHost = new Uri(authorityHost)
+                    });
+            HttpPipeline pipeline = HttpPipelineBuilder.Build(ClientOptions.Default, new BearerTokenAuthenticationPolicy(credential, "https://management.azure.com/.default"));
 
-            throw new InvalidOperationException("The environment has not become ready, check your TestEnvironment.IsEnvironmentReady scenario.");
+            // create the GET request for the resource group information
+            Request request = pipeline.CreateRequest();
+            Uri uri = new Uri($"{ResourceManagerUrl}/subscriptions/{subscription}/resourcegroups/{resourceGroup}?api-version=2021-04-01");
+            request.Uri.Reset(uri);
+            request.Method = RequestMethod.Get;
+
+            // send the GET request
+            Response response = await pipeline.SendRequestAsync(request, CancellationToken.None);
+
+            // resource group not found - nothing we can do here
+            if (response.Status == 404)
+            {
+                return;
+            }
+
+            // unexpected response => throw an exception
+            if (response.Status != 200)
+            {
+                throw await _clientDiagnostics.CreateRequestFailedExceptionAsync(response);
+            }
+
+            // parse the response
+            JsonElement json = JsonDocument.Parse(response.Content).RootElement;
+            if (json.TryGetProperty("tags", out JsonElement tags) && tags.TryGetProperty("DeleteAfter", out JsonElement deleteAfter))
+            {
+                DateTimeOffset deleteDto = DateTimeOffset.Parse(deleteAfter.GetString());
```

Comment:
This should be `TryParse`. It *should* be a correct `DateTimeOffset`, but someone could've messed with it incidentally.

Reply:
Well shouldn't they fix it then? It would make the delete resources script not work as well.

Reply:
Shouldn't who fix what? Anyone / anything could modify that metadata value.

The scripts that drive the live resource clean up effectively do try to cast using PowerShell's `-as`. Not as elegant, IMO, as `[DateTime]::TryParse()` in PowerShell but works to the same effect.

https://github.com/Azure/azure-sdk-tools/blob/360b3ae34a353e54488be50a8b1750ed5cdae0ec/eng/scripts/live-test-resource-cleanup.ps1#L67

Reply:
Whoever has an invalid value for the DeleteAfter tag should dig into why that is happening, and make it so they don't have invalid data.

Reply:
In other words, I don't see much value in hiding this because we probably want to know about it.

Reply:
Fair point. It is only isolated to this particular test run and, presumably, the RG would be cleaned up eventually. Though, @weshaggard, it does raise a good question (though I admit this is a corner case) whether RGs without `DeleteAfter` tags or with invalid `DeleteAfter` tags should raise an issue - not just logging we may not see in an otherwise successful run - but to actually aggregate and file issues. Since we're using AzDO, it shouldn't be hard to have it file issues automatically. I did have that setup up for an internal repo ages ago and don't remember much about the details other than there was a way.

Reply:
I agree with @jsquire that we don't necessary want to start failing on these resource groups unless we come up with some other way to opt-out of cleaning up these resources. Currently our clean-up automation just logs the resource groups without a DeleteAfter tag and haven't moved to do anything special there beyond some extra auditing.